src/Components/WebAssembly/Build/test/BuildIntegrationTests/BuildLazyLoadTest.cs

```diff
@@ -0,0 +1,272 @@
+</ItemGroup>");
+
+            var result = await MSBuildProcessManager.DotnetMSBuild(project, args: "/restore");
+
+            var buildOutputDirectory = project.BuildOutputDirectory;
+
+            Assert.FileExists(result, buildOutputDirectory, "wwwroot", "_framework", "blazor.boot.json");
+            Assert.FileExists(result, buildOutputDirectory, "wwwroot", "_framework", "_bin", "RclWithNoDeps.dll");
+
+            var bootJson = ReadBootJsonData(result, Path.Combine(buildOutputDirectory, "wwwroot", "_framework", "blazor.boot.json"));
+
+            var dynamicAssemblies = bootJson.resources.dynamicAssembly;
+            Assert.NotNull(dynamicAssemblies);
+            Assert.Contains("RclWithNoDeps.dll", dynamicAssemblies.Keys);
+        }
+
+        [Fact]
+        public async Task Build_LazyLoadRCLWithDependencies_Debug_Works()
+        {
+            // Arrange
+            using var project = ProjectDirectory.Create("standalone", additionalProjects: new[] { "razorclasslibrary", "rclwithpackages" });
+            project.Configuration = "Debug";
+
+            project.AddProjectFileContent(
+@"
+<ItemGroup>
+    <ProjectReference Include='..\rclwithpackages\rclwithpackages.csproj' BlazorLazyLoad='true'/>
+</ItemGroup>");
+
+            var result = await MSBuildProcessManager.DotnetMSBuild(project, args: "/restore");
+
+            var buildOutputDirectory = project.BuildOutputDirectory;
+
+            Assert.FileExists(result, buildOutputDirectory, "wwwroot", "_framework", "blazor.boot.json");
+            Assert.FileExists(result, buildOutputDirectory, "wwwroot", "_framework", "_bin", "rclwithpackages.dll");
+
+            var bootJson = ReadBootJsonData(result, Path.Combine(buildOutputDirectory, "wwwroot", "_framework", "blazor.boot.json"));
+
+            var dynamicAssemblies = bootJson.resources.dynamicAssembly;
+            Assert.NotNull(dynamicAssemblies);
+
+            // Packages that are dependencies of rclwithpackage and rclwithpackages itself are lazy-loaded
+            Assert.Contains("rclwithpackages.dll", dynamicAssemblies.Keys);
+            Assert.Contains("Newtonsoft.Json.dll", dynamicAssemblies.Keys);
+
+            // Dependencies of rclwithpackages that are also dependencies of
+            // non-lazy loaded components should not be lazy loaded
+            Assert.DoesNotContain("Microsoft.AspNetCore.Components.dll", dynamicAssemblies.Keys);
+        }
+
+        [Fact]
+        public async Task Build_LazyLoadPackage_Debug_Works()
+        {
+            // Arrange
+            using var project = ProjectDirectory.Create("standalone", additionalProjects: new[] { "razorclasslibrary" });
+            project.Configuration = "Debug";
+
+            project.AddProjectFileContent(
+@"
+<ItemGroup>
+    <PackageReference Include='Newtonsoft.Json' Version='12.0.3' BlazorLazyLoad='true'/>
+    <PackageReference Include='Polly' Version='7.2.1' BlazorLazyLoad='true' />
```

Comment:
If this works, great, but there's some possibility that the CI machines will refuse to restore arbitrary external packages. I guess we'll see :)

Reply:
`Newtonsoft.Json` should be OK to use in tests since it is included in our Dependencies files.

I think `Polly` will work either way but I've removed it, just in case.